### PR TITLE
KAFKA-15410: Expand partitions, segment deletion by retention and enable remote log on topic integration tests (1/4)

### PR DESCRIPTION
**Pull request by kamalcph**
1. PartitionsExpandTest
2. DeleteSegmentsByRetentionSizeTest
3. DeleteSegmentsByRetentionTimeTest and
4. EnableRemoteLogOnTopicTest

Enabled the test for both ZK and Kraft modes.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by kamalcph** [comment]
### PartitionsExpandTest Report

```scala

[SUCCESS] (1) create topic: Topic[name=topicA partition-count=1 replication-factor=2 segment-size=1 assignment={0=[0, 1]} properties={remote.storage.enable=true, local.retention.bytes=1, index.interval.bytes=1, segment.index.bytes=12}]

[SUCCESS] (2) produce-records: topicA-0
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k0, value=v0, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=0 record-count=1]
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=1 record-count=1]

[SUCCESS] (3) create-partitions: ExpandPartitionCountSpec[topicName=topicA, partitionCount=3, assignment={1=[0, 1], 2=[1, 0]}]

[SUCCESS] (4) consume-action:
  topic-partition = topicA-0
  fetch-offset = 0
  expected-record-count = 3
  expected-record-from-tiered-storage = 2

[SUCCESS] (5) expect-leader: topic-partition: topicA-1, replicaId: 0, electLeader: false

[SUCCESS] (6) expect-leader: topic-partition: topicA-2, replicaId: 1, electLeader: false

[SUCCESS] (7) produce-records: topicA-1
    ProducerRecord(topic=topicA, partition=1, headers=RecordHeaders(headers = [], isReadOnly = true), key=k0, value=v0, timestamp=null)
    ProducerRecord(topic=topicA, partition=1, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicA, partition=1, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    Segment[partition=topicA-1 offloaded-by-broker-id=0 base-offset=0 record-count=1]
    Segment[partition=topicA-1 offloaded-by-broker-id=0 base-offset=1 record-count=1]

[SUCCESS] (8) produce-records: topicA-2
    ProducerRecord(topic=topicA, partition=2, headers=RecordHeaders(headers = [], isReadOnly = true), key=k0, value=v0, timestamp=null)
    ProducerRecord(topic=topicA, partition=2, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicA, partition=2, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    Segment[partition=topicA-2 offloaded-by-broker-id=1 base-offset=0 record-count=1]
    Segment[partition=topicA-2 offloaded-by-broker-id=1 base-offset=1 record-count=1]

[SUCCESS] (9) produce-records: topicA-0
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k3, value=v3, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k4, value=v4, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k5, value=v5, timestamp=null)
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=3 record-count=1]
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=4 record-count=1]

[SUCCESS] (10) consume-action:
  topic-partition = topicA-0
  fetch-offset = 0
  expected-record-count = 6
  expected-record-from-tiered-storage = 5

[SUCCESS] (11) consume-action:
  topic-partition = topicA-1
  fetch-offset = 0
  expected-record-count = 3
  expected-record-from-tiered-storage = 2

[SUCCESS] (12) consume-action:
  topic-partition = topicA-2
  fetch-offset = 1
  expected-record-count = 2
  expected-record-from-tiered-storage = 1

Content of local tiered storage:

Broker IDFile                                       | Offsets |      Records 
----------------------------------------------------------------------------------------
topicA-2                                            |         |              
    00000000000000000001-TiygcgBrTDma-MXLyerzkQ.log |       1 |     (k1, v1) 
                                                    |         |              
    00000000000000000000-IY_ZJfQRSj2GbL4mjCZPdQ.log |       0 |     (k0, v0) 
                                                    |         |              
topicA-1                                            |         |              
    00000000000000000000-F7i40B5rQPmCmBjDAnsEHw.log |       0 |     (k0, v0) 
                                                    |         |              
    00000000000000000001-zVyeIx9WQ5mfd5tvDrW1cQ.log |       1 |     (k1, v1) 
                                                    |         |              
topicA-0                                            |         |              
    00000000000000000003-NArRT1ueTxCZeFkDP_PJIw.log |       3 |     (k3, v3) 
                                                    |         |              
    00000000000000000002-QZHcZK-SRh6i1JgZJCp8bA.log |       2 |     (k2, v2) 
                                                    |         |              
    00000000000000000004-r4ycZbBvTOSeQYpDybAm1w.log |       4 |     (k4, v4) 
                                                    |         |              
    00000000000000000000-RM48dExjQJGCESfEtNoD3g.log |       0 |     (k0, v0) 
                                                    |         |              
    00000000000000000001-ff2P80Q5Tm6142NN96EsfQ.log |       1 |     (k1, v1) 
                                                    |         |              
```

**Comment by kamalcph** [comment]
### DeleteTopicWithSecondaryStorageTest Report

```scala
[SUCCESS] (1) create topic: Topic[name=topicA partition-count=2 replication-factor=2 segment-size=1 assignment={0=[0, 1], 1=[1, 0]} properties={remote.storage.enable=true, local.retention.bytes=1, index.interval.bytes=1, segment.index.bytes=12}]

[SUCCESS] (2) produce-records: topicA-0
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k0, value=v0, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicA, partition=0, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=0 record-count=1]
    Segment[partition=topicA-0 offloaded-by-broker-id=0 base-offset=1 record-count=1]

[SUCCESS] (3) produce-records: topicA-1
    ProducerRecord(topic=topicA, partition=1, headers=RecordHeaders(headers = [], isReadOnly = true), key=k0, value=v0, timestamp=null)
    ProducerRecord(topic=topicA, partition=1, headers=RecordHeaders(headers = [], isReadOnly = true), key=k1, value=v1, timestamp=null)
    ProducerRecord(topic=topicA, partition=1, headers=RecordHeaders(headers = [], isReadOnly = true), key=k2, value=v2, timestamp=null)
    Segment[partition=topicA-1 offloaded-by-broker-id=1 base-offset=0 record-count=1]
    Segment[partition=topicA-1 offloaded-by-broker-id=1 base-offset=1 record-count=1]

[SUCCESS] (4) delete-topic: topicA
    RemoteDeleteSegment[source-broker-id=1 topic-partition=topicA-1 event-type=DELETE_SEGMENT event-count=2]
    RemoteDeleteSegment[source-broker-id=0 topic-partition=topicA-0 event-type=DELETE_SEGMENT event-count=2]

Content of local tiered storage:

Broker IDFile                                       | Offsets |      Records 
----------------------------------------------------------------------------------------
topicA-1                                            |         |              
topicA-0                                            |         |              

Condition not met within timeout 2000. Remote storage is not empty for topicA-0 ==> expected: <true> but was: <false>
Expected :true
Actual   :false
```

The test failed at action-5 where we expect the remote storage to be empty. Since, we don't differentiate the case between deleting remote log segments due to retention size/time/log-start-offset breach and topic deletion. The underlying remote-storage-manager have no way to remove any marker folders. In this case, topic-partition directory.

Will update the test to ignore empty partition directories.

**Comment by kamalcph** [comment]
@showuon @clolov @divijvaidya @satishd 

In this PR, added integration tests for four cases. Reassignment (shrink, move and expand) tests are failing as the leader returns local-log-start-offset instead of log-start-offset on reassignment, will fix the issue and cover it with integration test in the next PR.

Call for review. Please take a look.

File: storage/src/test/java/org/apache/kafka/tiered/storage/actions/ExpectEmptyRemoteStorageAction.java
Comment (kamalcph) [inline storage/src/test/java/org/apache/kafka/tiered/storage/actions/ExpectEmptyRemoteStorageAction.java]: We don't differentiate the case between segment deletion and topic deletion so the underlying remote-storage-manager doesn't know when to remove any topic-level marker files/folders. In case of LocalTieredStorage, there will be empty partition directories.

File: core/src/main/java/kafka/log/remote/RemoteLogManager.java
Comment (satishd) [inline core/src/main/java/kafka/log/remote/RemoteLogManager.java:304]: Good catch!

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/DeleteSegmentsByRetentionSizeTest.java
Comment (satishd) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/DeleteSegmentsByRetentionSizeTest.java]: Please add documentation of the test and the behaviour that will be easier for readers. This will be easy for folks who are new to this framework and trying to understand the test better.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java
Comment (satishd) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java:72]: It is good to prefix the key and values with partition info like "p0"/"p1"/"p2" so that the test verification checks for the respective messages in the partitions.

**Review by satishd — COMMENTED** [review]
Thanks @kamalcph for the PR. Left a couple of minor comments.

**Comment by clolov** [comment]
I will aim to provide a review by the end of the day! Thanks for the effort

File: core/src/main/scala/kafka/server/BrokerServer.scala
Comment (showuon) [inline core/src/main/scala/kafka/server/BrokerServer.scala:473]: Nice catch!

File: storage/src/test/java/org/apache/kafka/tiered/storage/actions/ExpectEmptyRemoteStorageAction.java
Comment (showuon) [reply storage/src/test/java/org/apache/kafka/tiered/storage/actions/ExpectEmptyRemoteStorageAction.java]: OK, so I think we should add all the explanation into the comment. From the current comment, it's not clear to me.

File: storage/src/test/java/org/apache/kafka/tiered/storage/utils/BrokerLocalStorage.java
Comment (showuon) [inline storage/src/test/java/org/apache/kafka/tiered/storage/utils/BrokerLocalStorage.java:59]: Thanks for the fix.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/DeleteSegmentsByRetentionTimeTest.java
Comment (showuon) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/DeleteSegmentsByRetentionTimeTest.java]: It's surprising the retentionTimeTest extends retentionSizeTest. Could we create an abstract `BaseDeleteSegmentsTest` class, and put the test spec there. Then we can have retentionTimeTest and retentionSizeTest tests extends the `BaseDeleteSegmentsTest`.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java
Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java:72]: Prefer to keep the format as it is. We can refactor it later if required.

**Review by showuon — APPROVED** [review]
LGTM!

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java
Comment (satishd) [reply storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java:72]: Sure, we can look into that later.

**Review by satishd — APPROVED** [review]
Thanks @kamalcph for addressing the review comments. LGTM.

**Comment by showuon** [comment]
Will wait for @clolov 's review before merging it.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/BaseDeleteSegmentsTest.java
Comment (clolov) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/BaseDeleteSegmentsTest.java:60]: For my understanding, why do we need this configuration update? How would it trigger the segment deletion?

**Review by clolov — APPROVED** [review]
Besides the small comment for my understanding the rest of the changes appear sensible. Thank you!

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/BaseDeleteSegmentsTest.java
Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/integration/BaseDeleteSegmentsTest.java:60]: default retention time is 7 days and retention bytes is unlimited so the config update is required.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java
Comment (divijvaidya) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java:86]: we risk forgetting this commented line. please add a JIRA

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/EnableRemoteLogOnTopicTest.java
Comment (divijvaidya) [inline storage/src/test/java/org/apache/kafka/tiered/storage/integration/EnableRemoteLogOnTopicTest.java:79]: why do we need the 5th argument here is we are anyways producing it below in produce() call? The produce action must already be adding these to "offloadable" no?

Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/integration/EnableRemoteLogOnTopicTest.java:79]: The 4th argument is the base-offset of the segment and the 5th argument is the contents of that segment. The test asserts that the segment contains only the expected messages in it. It is a `varargs` argument.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java
Comment (kamalcph) [reply storage/src/test/java/org/apache/kafka/tiered/storage/integration/PartitionsExpandTest.java:86]: Filed KAFKA-15431 to track this.

File: storage/src/test/java/org/apache/kafka/tiered/storage/integration/EnableRemoteLogOnTopicTest.java
Comment (divijvaidya) [reply storage/src/test/java/org/apache/kafka/tiered/storage/integration/EnableRemoteLogOnTopicTest.java:79]: ack

**Review by divijvaidya — APPROVED** [review]

**Comment by satishd** [comment]
There are a couple of unrelated tests failed in Jenkins jobs, merging to trunk and 3.6.

**Pull request merged**
